### ISSUE TITLE
Blob bindings  can't set ContentType and other properties

### ISSUE THREAD

**Issue by buchizo** [Azure/azure-webjobs-sdk-script#364]
The blob output binding feature set always application/octet-stream to ContentType property.
If possible, I would like to modified some blob properties.

And, Just now I don't found documents about some bindings.
Example, Blob output binding can use %rand-guid%. And file name placeholder can use {name} from input binding.

I want to know details about bindings.

**Comment by christopheranderson** [comment]
This is definitely an issue. It would be helpful to know if you're trying to use C#/Node. Could just share a code sample with us?

We have a variety of ideas for how to make this smoother. We'd appreciate your input on **how** you'd like to set the Content Type, as well.

**Comment by buchizo** [comment]
Example: https://gist.github.com/buchizo/2f8ac640a21cd78920eccbfd7e0d267d

If I use CloudBlockBlob, I can set ContentType property. But it's not easily.

**Comment by christopheranderson** [comment]
Agreed that you shouldn't have to use CloudBlockBlob - we should aim to make this work cross language, which means using generic types.

**Comment by christopheranderson** [comment]
We should think through how we'd design this.

**Comment by christopheranderson** [comment]
I think this is related to binding data improvements we've had in mind for node.

**Comment by mathewc** [comment]
Another related issue here: https://github.com/Azure/azure-webjobs-sdk-script/issues/465. In general, we want to allow "object" bindings. I.e., allow a user to specify `dataType: "object"` in their binding metadata, which we translate to a JObject binding on the extension, with the extension handling the mapping from Blob/EventData => JObject, which can then be exposed to script.

**Comment by Adamsimsy** [comment]
Any progress on being able to configure the content/mimetype when the output type is "blob" in Azure functions?

**Comment by mathewc** [comment]
Not yet - we'll get to it :)

**Comment by Adamsimsy** [comment]
Cool looking forward to it :D

**Comment by aantillonl** [comment]
I'd appreciate this feature too 👍

**Comment by ricklove** [comment]
Found this after making my feature request in #1664.

There, I specified that I would like to be able to set properties like this:

```json
{
    "name": "outOutputBlob",
    "type": "blob",
    "direction": "out",
    "path": "{container}/{blob}",
    "connection": "AZURE_STORAGE_CONNECTION_STRING",
    "properties": {
        "cacheControl": "public, max-age 86400",
        "contentType": "application/json",
        "contentEncoding": "gzip"
    }
}
```

**Comment by AlejandroGil** [comment]
Is there any progress on this?

**Comment by paulbatum** [comment]
No progress on this yet. Right now our main priority for JavaScript is getting the new out-of-proc model used in functions V2 to have feature and behavior parity with V1. Switching from Edge.js to a seperate node.exe process communicating over grpc is a pretty large change with a tail of bugs that will impact users trying to upgrade to V2 if not fixed.

**Comment by AlejandroGil** [comment]
Thank you Paul,

I finally resolved using Azure storage SDK:

```javascript
var azure = require('azure-storage');
var streamifier = require('streamifier');

var options = {contentSettings:{contentType:'application/xml'}}
content_stream = streamifier.createReadStream("content");
var blobService = azure.createBlobService("DefaultEndpointsProtocol=https;AccountName=.....");
content_stream.pipe(blobService.createWriteStreamToBlockBlob(container, outputFileName, options, function(error, result, response){...}
```

**Comment by lucasmike** [comment]
any update on this issues?

**Comment by paulbatum** [comment]
My comment above still reflects the current state.

**Comment by nemanjamil** [comment]
This is one of the solutions for upload image and set the adequate mime - content type
https://github.com/nemanjamil/azure-functions/blob/master/azure_upload_image.js

**Comment by mikhailshilkov** [comment]
This basically means that [the example from the docs](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-blob#input---javascript-example) ([repeated twice](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-blob#output---javascript-example)) doesn't work. For some files (e.g. text/html), it copies the contents but breaks the content type. For others (e.g. png), it breaks the content too: I got the encoding messed up and a "copy" file is twice as long.

Tested with this bindings:

```
{"disabled":false,"bindings":[
  {"authLevel":"anonymous","type":"httpTrigger","direction":"in","name":"req","route":"{name}"},
  {"type":"http","direction":"out","name":"response"},
  {"name":"original","type":"blob","direction":"in","path":"files/{name}","connection":"StorageConnectionStringKey"},
  {"name":"copy","type":"blob","direction":"out","path":"files/copy-{name}","connection":"StorageConnectionStringKey"}
]}
```

**Comment by forki** [comment]
still not possible? that's a bummer

**Comment by shahedurrahman** [comment]
Is there any progress on this? Thanks

**Comment by eastonharvey** [comment]
Any updates on this? Would like to set properties on a Service Bus binding.

**Comment by matses** [comment]
any updates on this?

**Comment by alexanderomnix** [comment]
any update?

**Comment by SashiDareddy** [comment]
2021 and still waiting for this feature.

**Comment by invisibleaxm** [comment]
Any chance you we can get an update?

**Comment by NewFuture** [comment]
**5 Years** have passed!

**Comment by shanselman** [comment]
👀

**Comment by WestDiscGolf** [comment]
Just came across this issue. It also applied to the new isolated process model for C# as well. Any news?

**Comment by shanselman** [comment]
This is how Jeff Hollan solved it for me
https://github.com/shanselman/azurefridayaggregator/pull/1/files

It's not clean, but it works. I would prefer that I be able to set the
content type more readily in a single line.

**Comment by WestDiscGolf** [comment]
@shanselman thanks for sharing the link. The Blob bindings in the new isolated model don't give those options unfortunately. I guess the only option will be to inject a `BlockBlobClient` manually through DI and save it that way instead of an `BlobOutput` binding.

**Comment by WestDiscGolf** [comment]
What I ended up doing was not using the binding at all.

1) Registering the BlobServiceClient in `ConfigureServices`

```
services.AddAzureClients(builder =>
        {
            builder.AddBlobServiceClient(hostContext.Configuration["AzureWebJobsStorage"]);
        });
```

2. Injecting into the azure function, through dependency injection, a `BlobServiceClient`.

3. In the timer function method getting a Blob Container Client.

```
var containerClient = _blobServiceClient.GetBlobContainerClient("my-blob-name");
```

4. Creating a BlockBlobClient from the Container client.
```
var blockClient = containerClient.GetBlockBlobClient("blob-name.json");
```

5. Writing the data to a stream and uploading it like @shanselman example.
```
Stream jsonStream = new MemoryStream();
await JsonSerializer.SerializeAsync(jsonStream, mydata, new JsonSerializerOptions() { WriteIndented = true });
jsonStream.Position = 0;
await blockClient.UploadAsync(jsonStream, new BlobHttpHeaders() { ContentType = "application/json" });
```

Hope this helps others.

**Comment by Guzzter** [comment]
Thanks @shanselman for the always entertaining Youtube videos, this time about his [Azure Friday Web API issue](https://www.youtube.com/watch?v=u9vHV_KpTC8). Also thanks for the helpful pointers @WestDiscGolf!

I managed to brew my own Azure Function storing files with filename generation at runtime with Json content type. If you want to see how I did it: [https://github.com/Guzzter/RatingViewerToJson](https://github.com/Guzzter/RatingViewerToJson)

**Comment by elrondfeng** [comment]
is there a solution to this problem?

**Comment by gamecubate** [comment]
And this is  why we switched to some other platform.

**Comment by mk7exe** [comment]
Seriously! This is still an issue?! Wow! No update?

**Comment by RoryMcCrossan** [comment]
Are you actually joking... 8 YEARS LATER and this has still not been adressed? The entire Azure blob storage offering is an inconsistent mess. 

This is unbelieveable. My bar of expectation for you guys is on the floor, and here you are bringing shovels.
